src/portable/synopsys/dwc2/dwc2_esp32.h

```diff
@@ -33,37 +33,73 @@
 #endif
 
 #include "esp_intr_alloc.h"
-#include "soc/periph_defs.h"
-//#include "soc/usb_periph.h"
+#include "soc/interrupts.h"
+
+#if CFG_TUSB_MCU == OPT_MCU_ESP32P4
+  #define DWC2_PERIPH_COUNT    2
+  #define DWC2_HS_PERIPH_BASE  0x50000000UL
+  #define DWC2_FS_PERIPH_BASE  0x50004000UL
+  #define DWC2_HS_INTR_SOURCE  ETS_USB_OTG_INTR_SOURCE
+  #define DWC2_FS_INTR_SOURCE  ETS_USB_OTG_INTR_SOURCE // TODO: change to the correct interrupt source
+#elif TU_CHECK_MCU(OPT_MCU_ESP32S2, OPT_MCU_ESP32S3)
+  #define DWC2_PERIPH_COUNT    1
+  #define DWC2_FS_PERIPH_BASE  0x60080000UL
+  #define DWC2_FS_INTR_SOURCE  ETS_USB_INTR_SOURCE
+#else
+  #error "DWC2_REG_BASE not defined for current MCU"
+#endif
 
-#define DWC2_REG_BASE       0x60080000UL
-#define DWC2_EP_MAX         6             // USB_OUT_EP_NUM. TODO ESP32Sx only has 5 tx fifo (5 endpoint IN)
+#if DWC2_HS_PERIPH_BASE
+  #define DWC2_HS_EP_MAX       15             // USB_OUT_EP_NUM
+  #define DWC2_HS_EP_FIFO_SIZE 4096
+#endif
 
+#if DWC2_FS_PERIPH_BASE
+  #define DWC2_FS_EP_MAX       6             // USB_OUT_EP_NUM. TODO ESP32Sx only has 5 tx fifo (5 endpoint IN)
+  #define DWC2_FS_EP_FIFO_SIZE 1024
+#endif
+
+// OTG HS always has higher number of endpoints than FS
+// Declare the DWC2_EP_MAX as a higher number, because dcd_dwc2.c has a xfer_status
+// buffer, tied to the EP MAX value.
+// IMPORTANT: Be careful during usage of two USB periphs (HS and FS) at the same time with full EPs load.
+#ifdef USB_OTG_HS_PERIPH_BASE
+  #define DWC2_EP_MAX   DWC2_HS_EP_MAX
+#else
+  #define DWC2_EP_MAX   DWC2_FS_EP_MAX
+#endif
+
+// On ESP32 for consistency we associate
+// - Port0 to OTG_FS, and Port1 to OTG_HS
 static const dwc2_controller_t _dwc2_controller[] =
 {
-  { .reg_base = DWC2_REG_BASE, .irqnum = 0, .ep_count = DWC2_EP_MAX, .ep_fifo_size = 1024 }
+#ifdef DWC2_FS_PERIPH_BASE
+  { .reg_base = DWC2_FS_PERIPH_BASE, .irqnum = DWC2_FS_INTR_SOURCE, .ep_count = DWC2_FS_EP_MAX, .ep_fifo_size = DWC2_FS_EP_FIFO_SIZE },
+#endif
+
+#ifdef DWC2_HS_PERIPH_BASE
+  { .reg_base = DWC2_HS_PERIPH_BASE, .irqnum = DWC2_HS_INTR_SOURCE, .ep_count = DWC2_HS_EP_MAX, .ep_fifo_size = DWC2_HS_EP_FIFO_SIZE },
+#endif
 };
 
-static intr_handle_t usb_ih;
+static intr_handle_t usb_ih[DWC2_PERIPH_COUNT];
```

Comment:
I don't like the way how `usb_ih` is using (based on `rhport`). 
I would prefer to add a verification for `rhport` argument to exclude the possibility to use `rhport >= DWC_PERIPH_COUNT`. (and do not get the out of boundary access for `usb_ih[]`) 

@tore-espressif, @leeebo 
what do you think?

Reply:
I had a quick look and probably there is no need to change this, as this variable is not exposed to users